v3/mathjax3-ts/core/MathDocument.ts

```diff
@@ -0,0 +1,401 @@
+        if (!Array.isArray(this.inputJax)) {
+            this.inputJax = [this.inputJax];
+        }
+    }
+
+    /*
+     * @return{string}  The kind of document
+     */
+    public get kind() {
+        return (this.constructor as MathDocumentClass).KIND;
+    }
+
+    /*
+     * @override
+     */
+    public findMath(options: OptionList) {
+        this.processed.findMath = true;
+        return this;
+    }
+
+    /*
+     * @override
+     */
+    public compile() {
+        if (!this.processed.compile) {
+            for (const math of this.math) {
+                math.compile(this);
+            }
+            this.processed.compile = true;
+        }
+        return this;
+    }
+
+    /*
+     * @override
+     */
+    public typeset() {
+        if (!this.processed.typeset) {
+            for (const math of this.math) {
+                math.typeset(this);
+            }
+            this.processed.typeset = true;
+        }
+        return this;
+    }
+
+    /*
+     * @override
+     */
+    public getMetrics() {
+        if (!this.processed.getMetrics) {
+            this.outputJax.getMetrics(this);
+            this.processed.getMetrics = true;
+        }
+        return this;
+    }
+
+    /*
+     * @override
+     */
+    public addEventHandlers() {
+        this.processed.addEventHandlers = true;
+        return this;
+    }
+
+    /*
+     * @override
+     */
+    public updateDocument() {
+        if (!this.processed.updateDocument) {
+            for (const math of this.math.reversed()) {
+                math.updateDocument(this);
+            }
+            this.processed.updateDocument = true;
+        }
+        return this;
+    }
+
+    /*
+     * @override
+     */
+    public removeFromDocument(restore: boolean = false) {
+        return this;
+    }
+
+    /*
+     * @override
+     */
+    public state(state: number, restore: boolean = false) {
+        for (const math of this.math) {
+            math.state(state, restore);
+        }
+        if (state < STATE.INSERTED) {
+            this.processed.updateDocument = false;
+        }
+        if (state < STATE.TYPESET) {
+            this.processed.typeset = false;
+            this.processed.addEventHandlers = false;
+            this.processed.getMetrics = false;
+        }
+        if (state < STATE.COMPILED) {
+            this.processed.compile = false;
+        }
+        return this;
+    }
+
+    /*
+     * @override
+     */
+    public reset() {
+        for (const key of Object.keys(this.processed)) {
+            this.processed[key] = false;
+        }
+        return this;
+    }
+
+    /*
+     * @override
+     */
+    public clear() {
+        this.reset();
+        this.math.clear();
+        return this;
+    }
+
+    /*
+     * @override
+     */
+    public concat(list: MathList) {
+        this.math.merge(list);
+        return this;
+    }
+
+}
+
+let STATE = AbstractMathDocument.STATE;
```

Comment:
Left over from pre-Typescript implementation.  ES6 doesn't handle static variables in the same way Typescript does.